File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
from collections import deque

def solution(s):
    dq = deque(s)
    answer = 0

    if len(s) % 2 != 0:
        return 0

    for i in range(len(s)):
        if check_func(dq):
            answer += 1
        
        dq.append(dq.popleft())

    return answer

def check_func(dq):
    stack = []
    pair = {')': '(', ']': '[', '}': '{'}

    for ch in dq:
        if ch in '([{':
            stack.append(ch)
        else:
            if not stack or stack[-1] != pair[ch]:
                return False
            stack.pop()

    return not stack

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    public int solution(String s) {
        int answer = 0;
        HashMap<Character, Character> map = new HashMap<>();
        
        map.put(')', '(');
        map.put('}', '{');
        map.put(']', '[');
       
        int length = s.length();
         s += s;
        
        A:for(int i=0; i<length; i++) {
            ArrayDeque<Character> stack = new ArrayDeque<>();
            for(int j = i; j< i+length; j++) {
                char c = s.charAt(j);
                if(!map.containsKey(c)) {
                    stack.push(c);
                } else if(stack.isEmpty() || !stack.pop().equals(map.get(c))){
                    continue A;
                }
            }
            if(stack.isEmpty()) {
                answer += 1;
            }
        }
       
        return answer;
    }
}